src/game/WorldSession.cpp

```diff
@@ -102,17 +99,16 @@ WorldSession::~WorldSession()
         LogoutPlayer(true);
 
     /// - If have unclosed socket, close it
-    if (m_Socket)
-    {
-        m_Socket->CloseSocket();
-        m_Socket->RemoveReference();
-        m_Socket = nullptr;
-    }
+    if (!m_Socket->IsClosed())
+        m_Socket->Close();
 
     ///- empty incoming packet queue
-    WorldPacket* packet = nullptr;
-    while (_recvQueue.next(packet))
+    while (!m_packetQueue.empty())
```

Comment:
I don't think it solves the problem completely.

The network thread reads data en calls the function OnRead.
This function will then call ProcessIncomingData:
https://github.com/namreeb/mangos-wotlk/blob/boost.asio/src/game/WorldSocket.cpp#L140

In this function you check if the socket is already closed:
https://github.com/namreeb/mangos-wotlk/blob/boost.asio/src/game/WorldSocket.cpp#L197

However if you close it just after this check we end up calling QueuePacket (of the WorldSession):
https://github.com/namreeb/mangos-wotlk/blob/boost.asio/src/game/WorldSocket.cpp#L240

The world session might however already have been deleted resulting in a crash.